File: simulatedExchange.cpp
#include <iostream>
#include <queue>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Exchange {
public:

       priority_queue< pair< pair<int, int>,int> > sellOrder; //<<price,uid>,volume>
       priority_queue< pair< pair<int, int>,int> > buyOrder;
       unordered_map<int,pair<int, double>> orderBook; // {id: (volume,avgPrice)}
       int currId;

       Exchange(){
              this->currId = 1;
       }

       // InputOrder receives order, and return assigned order Id.
       // Parameters:
       //   int side:   0 is buy, 1 is sell
       //   int volume: order's quantity
       //   int price:  order's price
       // Return:
       //   int: order Id
       int inputOrder(int side, int volume, int price){
              int uid = currId;
              currId += 1;
              orderBook[uid] = make_pair(0,0.00); // tradeVol, avgPrice
              // cout<<orderBook[uid].second<<endl;
              if(side>0)
                     sellOrder.push(make_pair(make_pair(-1*price,-1*uid),volume));
              else
                     buyOrder.push(make_pair(make_pair(price,-1*uid),volume));

              matchOrder();
              return uid;
       };
       

       void matchOrder(){
              while(!sellOrder.empty() && !buyOrder.empty()){
                     pair< pair<int, int>,int> buy = buyOrder.top();
                     pair< pair<int, int>,int> sell = sellOrder.top(); 
                     int p1,id1,v1,p2,id2,v2;
                     p1 = buy.first.first; p2 = -1*(sell.first.first);
                     id1 = -1*(buy.first.second); id2 = -1*(sell.first.second);
                     v1 = buy.second; v2 = sell.second;

                     buyOrder.pop();
                     sellOrder.pop();

                     if(p2>p1) return;

                     // cout<<"processing: "<<id1<<" & "<<id2<<endl;
                     if(v1==v2){
                            orderBook[id1].second = (orderBook[id1].second*orderBook[id1].first + 1.0*p2*v1)/(1.0*(v1+orderBook[id1].first));
                            orderBook[id2].second = (orderBook[id2].second*orderBook[id2].first + 1.0*p1*v2)/(1.0*(v2+orderBook[id2].first));
                            orderBook[id1].first += v1;
                            orderBook[id2].first += v2;
                     }
                            
                     else if(v1>v2){
                            orderBook[id1].second = 0.0+ (orderBook[id1].second*orderBook[id1].first + p2*v2)/(1.0*(v2+orderBook[id1].first ));
                            orderBook[id2].second = 0.0+ (orderBook[id2].second*orderBook[id2].first + p1*v2)/(1.0*(v2+orderBook[id2].first));
                            orderBook[id1].first += v2;
                            orderBook[id2].first += v2;
                            buyOrder.push(make_pair(make_pair(p1,-1*id1),v1-v2));

                     }
                            
                     else{ //v2>v1
                            orderBook[id1].second = 0.0+(orderBook[id1].second*orderBook[id1].first + p2*v1)/(1.0*(v1+orderBook[id1].first));
                            orderBook[id2].second = 0.0+(orderBook[id2].second*orderBook[id2].first + p1*v1)/(1.0*(v1+orderBook[id2].first));
                            orderBook[id1].first += v1;
                            orderBook[id2].first += v1;
                            sellOrder.push(make_pair(make_pair(-1*p2,-1*id2),v2-v1));

                     } 
              }

       return;
       };


       // QueryOrderTrade queries order's trade volume and average price.
       // Parameters:
       //   int orderId: assigned order Id
       //   int averagePrice: return order's average trade price
       // Return:
       //   int: return order's trade volume
       void showOrderBook(){
              while(!buyOrder.empty()){
                     cout<<buyOrder.top().first.first<<" "<<buyOrder.top().first.second<<" "<<buyOrder.top().second <<endl;
                     buyOrder.pop();
              }
              cout<<endl;
              while(!sellOrder.empty()){
                     cout<<sellOrder.top().first.first<<" "<<sellOrder.top().first.second<<" "<<sellOrder.top().second <<endl;
                     sellOrder.pop();
              }
       };

       int QueryOrderTrade(int orderId, double& averagePrice){
              pair<int,double> p = orderBook[orderId];
              averagePrice = p.second;
              int tradeVolume = p.first;
              return tradeVolume;
       };

 };

int main() {
       Exchange ex;
       std::vector<int> orders;
       orders.push_back(ex.inputOrder(0, 1, 100));
       orders.push_back(ex.inputOrder(0, 2, 101));
       orders.push_back(ex.inputOrder(0, 3, 102));
       orders.push_back(ex.inputOrder(1, 4, 100));
       orders.push_back(ex.inputOrder(1, 5, 101));
       orders.push_back(ex.inputOrder(1, 6, 102));
       // ex.showOrderBook();
       for (auto& orderId : orders)
       {
              double averagePrice;
              int tradeVolume = ex.QueryOrderTrade(orderId, averagePrice);
              std::cout << "orderId: " << orderId << " tradeVolume: " << tradeVolume << " averagePrice: " << averagePrice << std::endl;
       }
               
       return 0; 
}

File: simulatedExchange.py
from collections import deque
import heapq

class Exchange:
    def __init__(self):
        self.orderBook = {}
        # use heaps
        self.buyOrder = [] # store buy in decreasing order of price, increasing order of time
        self.sellOrder = [] # store sell in increasing order of price, increasing order of time
        self.currId = 1
        pass

    
    def matchOrder(self):
        while self.buyOrder and self.sellOrder:
            buy,sell = heapq.heappop(self.buyOrder),heapq.heappop(self.sellOrder)
            p1,id1,v1 = -buy[0][0],buy[0][1],buy[1] # p1 is maximum price, a buyer is willing to buy 
            p2,id2,v2 = sell[0][0],sell[0][1],sell[1] # p2 is minimum price, a seller is willing to sell
            
            if p2>p1: return
            #print("processing: ",id1," & ",id2)
            if v1==v2:
                self.orderBook[id1][-1] = (self.orderBook[id1][-1]*self.orderBook[id1][-2] + p2*v1)/(v1+self.orderBook[id1][-2])
                self.orderBook[id2][-1] = (self.orderBook[id2][-1]*self.orderBook[id2][-2] + p1*v2)/(v2+self.orderBook[id2][-2])
                self.orderBook[id1][-2]+=v1
                self.orderBook[id2][-2]+=v2
            elif v1>v2:
                self.orderBook[id1][-1] = (self.orderBook[id1][-1]*self.orderBook[id1][-2] + p2*v2)/(v2+self.orderBook[id1][-2])
                self.orderBook[id2][-1] = (self.orderBook[id2][-1]*self.orderBook[id2][-2] + p1*v2)/(v2+self.orderBook[id2][-2])
                self.orderBook[id1][-2]+=v2
                self.orderBook[id2][-2]+=v2
                heapq.heappush(self.buyOrder,[(-p1,id1),v1-v2])
            else: # v2>v1
                self.orderBook[id1][-1] = (self.orderBook[id1][-1]*self.orderBook[id1][-2] + p2*v1)/(v1+self.orderBook[id1][-2])
                self.orderBook[id2][-1] = (self.orderBook[id2][-1]*self.orderBook[id2][-2] + p1*v1)/(v1+self.orderBook[id2][-2])
                self.orderBook[id1][-2]+=v1
                self.orderBook[id2][-2]+=v1
                heapq.heappush(self.sellOrder,[(p2,id2),v2-v1])
        
        return
        
    
    def InputOrder(self, side, volume, price):
        """
        InputOrder receives order, and return assigned order Id.
        :param side: 0 is buy, 1 is sell,-1 means not valid
        :param volume:order's quantity
        :param price:order's prices
        :return: order Id, an integer
        """
        if side<0: return None
        uid = self.currId
        self.currId += 1
        self.orderBook[uid]=[0,0] # total Vol, avg Price
        
        if side>0: #sell
            heapq.heappush(self.sellOrder,[(price, uid), volume])
        else: #buy
            heapq.heappush(self.buyOrder,[(-price, uid), volume])
        
        self.matchOrder()
        
        return uid

    def QueryOrderTrade(self, orderId):
        """
        queries order's trade volume and average price.
        :param orderId: assigned order Id
        :return: (order's trade volume, avg_price),tuple
        """        
        tradeVol, avg_price = self.orderBook[orderId][-2],self.orderBook[orderId][-1]
        return (tradeVol, avg_price)
    
    def printOrderBook(self):
        print(self.orderBook)

    

if __name__ == "__main__":
    # 0 is buy, 1 is sell
    ex = Exchange()
    orders = list()
    orders.append(ex.InputOrder(0, 1, 100))
    orders.append(ex.InputOrder(0, 2, 101))
    orders.append(ex.InputOrder(0, 3, 102))
    orders.append(ex.InputOrder(1, 4, 100))
    orders.append(ex.InputOrder(1, 5, 101))
    orders.append(ex.InputOrder(1, 6, 102))
    
    # print(ex.sellOrder)
    # print(ex.buyOrder)
    # ex.printOrderBook()
    
    for orderId in orders:
        tradeVolume, avgPrice = ex.QueryOrderTrade(orderId)
        print("orderId:%d, tradeVolume:%d, averagePrices:%f"%(orderId,tradeVolume,avgPrice))

